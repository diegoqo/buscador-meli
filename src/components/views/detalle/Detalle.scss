.detalle {
  background-color: #ededed;
  display: flex;
  flex-direction: row;
  justify-content: center;
  &__container-producto {
    background-color: #ffffff;
    width: 80% !important;
  }
  &__container-imagen {
    padding: 10px 0 10px 0 !important;
    &--item {
      box-shadow: none !important;
    }
    &--detalle {
      width: 80%;
    }
    &--descripcion {
      text-align: left;
      padding: 13px;
      p {
        word-break: break-all;
      }
    }
  }
  &__container-precio {
    padding: 10px 0 10px 0 !important;
    &--item-comprar {
      box-shadow: none !important;
    }
    &--comprar {
      text-align: left;
      margin: 5%;
    }
    &--button {
      text-align: center;
    }
  }
}

.container-miga {
  margin: 16px;
}
