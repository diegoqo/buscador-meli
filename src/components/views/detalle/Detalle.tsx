import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import CajaBusqueda from '../../common/cajabusqueda/CajaBusqueda';
import { ICategorias, IProductos } from '../../../modelo/interfaces';
import MigaDePan from '../../common/migadepan/MigaDePan';
import { Box, Button, Grid, Paper, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import './Detalle.scss'
import useFetchData from '../../../hooks/useFetchData';
import { ConstructorProductoDetalle1, ConstructorProductoDetalle2 } from '../../../utils/ConstructoresInterfaces';
import { FormatearPrecio } from '../../../utils/Formateador';
import Helmet from 'react-helmet';

const R = require('ramda');

const Item = styled(Paper)(({theme}) => ({
    backgroundColor: theme.palette.mode === 'dark' ? '#1A2027' : '#fff',
    ...theme.typography.body2,
    padding: theme.spacing(1),
    textAlign: 'center',
    color: theme.palette.text.secondary,
}));

const Detalle = () => {
    const {id} = useParams();
    const [categorias, setCategorias] = useState<ICategorias[]>([]);
    const [productoDetalle1, setProductoDetalle1] = useState<IProductos>();
    const [productoDetalle, setProductoDetalle] = useState<IProductos>();
    const urlServer = process.env.REACT_APP_API_URL;

    const {
        response: responseDetalle1
    } = useFetchData(`${urlServer}/search-item`, {id: id}, R.pathOr('', [], id));

    const {
        response: responseDetalle2
    } = useFetchData(`${urlServer}/search-description`, {id: id}, R.pathOr('', [], id));

    useEffect(() => {
        if (responseDetalle1) {
            const detalle1: IProductos = ConstructorProductoDetalle1(responseDetalle1);
            setProductoDetalle1(detalle1);
        }

    }, [responseDetalle1]);

    useEffect(() => {
        const productoConDescripcion: IProductos = ConstructorProductoDetalle2(R.pathOr([], [], productoDetalle1), R.pathOr([], [], responseDetalle2));
        const categoriasLocalStorage = localStorage.getItem('categorias');
        const categoriasLst: ICategorias[] = R.pathOr([], [], categoriasLocalStorage).length > 0
            ? Array.from(JSON.parse(R.pathOr([], [], categoriasLocalStorage)))
            : [];
        const categoriaEncontrada = R.find(R.propEq('id', productoConDescripcion?.item?.category))(categoriasLst);

        if (categoriaEncontrada) {
            setCategorias([categoriaEncontrada, {
                id: R.pathOr('', ['item', 'id'], productoConDescripcion),
                nombre: R.pathOr('', ['item', 'title'], productoConDescripcion),
            }]);
        }
        setProductoDetalle(productoConDescripcion);
    }, [productoDetalle1]);


    return (
        <><Helmet>
            <title>Detalle</title>
            <meta name="description" content="Página de detalle de producto" />
            <meta name="robots" content="INDEX,FOLLOW" />
        </Helmet>
            <Box>
                <CajaBusqueda/>
                {
                    productoDetalle !== undefined &&
                    <>
                        {categorias?.length > 0 &&
                            <Box className={'container-miga'}><MigaDePan categorias={categorias}/></Box>}
                        <Box className={'container-producto'}>
                            <Grid container spacing={2}>
                                <Grid item xs={7}>
                                    <Item><img className={'img-detalle'}
                                               src={productoDetalle?.item?.picture}/>
                                        <Box className={'container-descripcion'}>
                                            <h1>Descripción del producto</h1>
                                            <Box>
                                                <Typography variant="body2" color="text.secondary"
                                                            key={`${productoDetalle.item?.id}-grid-typo-shipp`}>
                                                    {productoDetalle.item?.description}
                                                </Typography>
                                            </Box>
                                        </Box>
                                    </Item>

                                </Grid>
                                <Grid item xs>
                                    <Item className={'container-item-comprar'}>
                                        <Box className={'container-comprar'}>
                                            <Box
                                                className={'container-items-comprar'}>{`${R.pathOr('Condición no disponible', ['item', 'condition'], productoDetalle)} - ${R.pathOr('No disponible número de', ['item', 'sold_quantity'], productoDetalle)} vendidos`}</Box>
                                            <Box className={'container-items-comprar'}>
                                                <h3>{R.pathOr('Título no disponible', ['item', 'title'], productoDetalle)}</h3></Box>
                                            <Box className={'container-items-comprar'}>
                                                <h1>{R.pathOr(undefined, ['item', 'price'], productoDetalle) !== undefined
                                                    && FormatearPrecio(R.pathOr('ARS', ['item', 'price', 'curremcy'], productoDetalle),
                                                        R.pathOr(0, ['item', 'price', 'amount'], productoDetalle))}
                                                </h1>
                                            </Box>
                                            <Box className={'container-button-comprar'}><Button variant="contained"
                                                                                    onClick={() => alert('Todo: Carrito de compras :)')}>Comprar</Button></Box>
                                        </Box>
                                    </Item>
                                </Grid>
                            </Grid>
                        </Box>
                    </>
                }
            </Box>
        </>

    );
}


export default Detalle;
