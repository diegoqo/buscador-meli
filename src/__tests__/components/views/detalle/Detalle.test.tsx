import { cleanup, render } from '@testing-library/react';
import Detalle from '../../../../components/views/detalle/Detalle';
import { BrowserRouter } from 'react-router-dom';

afterEach(cleanup);

describe('Probando el componente Detalle', () => {

    it('deberÃ­a mostrar Detalle correctamente', () => {
        const { asFragment } = render(<BrowserRouter><Detalle/></BrowserRouter>);
        expect(asFragment()).toMatchSnapshot();
    })
})
