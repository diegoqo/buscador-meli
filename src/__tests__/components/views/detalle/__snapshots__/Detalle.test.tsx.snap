// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente Detalle debería mostrar Detalle correctamente 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="container-caja-busqueda MuiBox-root css-0"
    >
      <div
        class="container-img-buscador MuiBox-root css-0"
      >
        <img
          src="https://http2.mlstatic.com/frontend-assets/ui-navigation/5.18.8/mercadolibre/logo__small@2x.png"
        />
      </div>
      <div
        class="container-input-buscador MuiBox-root css-0"
      >
        <form
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 paper-input-buscador css-1ps6pg7-MuiPaper-root"
        >
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary input-buscador css-1le28he-MuiInputBase-root"
            data-testid="input-busqueda"
          >
            <input
              class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
              placeholder="Buscar productos, marcas y más..."
              type="text"
              value=""
            />
          </div>
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1gtpq4r-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </div>
    <div
      class="container-producto MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 css-uqwprf-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1633um2-MuiPaper-root"
          >
            <img
              class="img-detalle"
            />
            <div
              class="container-descripcion MuiBox-root css-0"
            >
              <h1>
                Descripción del producto
              </h1>
              <div
                class="MuiBox-root css-0"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  No hay descripción para el producto
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 container-item-comprar css-1633um2-MuiPaper-root"
          >
            <div
              class="container-comprar MuiBox-root css-0"
            >
              <div
                class="container-items-comprar MuiBox-root css-0"
              >
                Condición no disponible - No disponible número de vendidos
              </div>
              <div
                class="container-items-comprar MuiBox-root css-0"
              >
                <h3>
                  Título no disponible
                </h3>
              </div>
              <div
                class="container-items-comprar MuiBox-root css-0"
              >
                <h1 />
              </div>
              <div
                class="container-button-comprar MuiBox-root css-0"
              >
                <button
                  class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Comprar
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
